File: sparktk-core/src/main/scala/org/trustedanalytics/sparktk/graph/internal/ops/ConnectedComponents.scala
/**
 *  Copyright (c) 2016 Intel Corporation 
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package org.trustedanalytics.sparktk.graph.internal.ops

import org.trustedanalytics.sparktk.frame.Frame
import org.apache.spark.sql.functions._
import org.graphframes.GraphFrame
import org.graphframes.GraphFrame.ID
import org.apache.spark.sql.DataFrame
import org.graphframes.lib.AggregateMessages

import org.apache.spark.sql.Column
import org.apache.spark.sql.functions.lit

import org.trustedanalytics.sparktk.graph.internal.{ GraphState, GraphSummarization, BaseGraph }

trait ConnectedComponentsSummarization extends BaseGraph {
  /**
   *
   * Connected components are disjoint subgraphs in which all vertices are
   * connected to all other vertices in the same component via paths, but not
   * connected via paths to vertices in any other component.
   *
   * @return The dataframe containing the vertices and which component they belong to
   */
  def connectedComponents(): Frame = {
    execute[Frame](ConnectedComponents())
  }
}

case class ConnectedComponents() extends GraphSummarization[Frame] {

  val index = "component"
  val indexNew = "componentNew"

  override def work(state: GraphState): Frame = {

    var graph = GraphFrame(state.graphFrame.vertices.withColumn(index, monotonicallyIncreasingId()), state.graphFrame.edges)
    var continue = true
    var getMin = udf { (a: Long, b: Long) => if (a > b) b else a }

    while (continue) {
      val updatedComponent = graph
        .aggregateMessages
        .sendToDst(getMin(AggregateMessages.src(index), AggregateMessages.dst(index)))
        .sendToSrc(getMin(AggregateMessages.src(index), AggregateMessages.dst(index)))
        .agg(min(AggregateMessages.msg).as(indexNew))

      val joinedComponent = updatedComponent
        .join(graph.vertices, graph.vertices(ID) === updatedComponent(ID))
        .drop(graph.vertices(ID))

      // TODO: switch this to an accumulator argument
      continue = joinedComponent.where(col(index) !== col(indexNew)).count() > 0

      val newVertices = joinedComponent
        .drop(index)
        .withColumnRenamed(indexNew, index)

      val unCachedVertices = AggregateMessages.getCachedDataFrame(newVertices)

      graph = GraphFrame(unCachedVertices, state.graphFrame.edges)
    }

    new Frame(graph.vertices)
  }
}

File: python/sparktk/graph/ops/connected_components.py
# vim: set encoding=utf-8

#  Copyright (c) 2016 Intel Corporation 
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
#


def connected_components(self):
    """

    Connected components determines groups all the vertices in a particular graph
    by whether or not there is path between these vertices. This method returns
    a frame with the vertices and their corresponding component

    Parameters
    ----------

    :return: (Frame) Frame containing the vertex id's and their components

    Examples
    --------

        >>> vertex_schema = [('id', int)]
        >>> edge_schema = [('src', int), ('dst', int)]

        >>> vertex_rows = [ [1], [2], [3], [4], [5] ]
        >>> edge_rows = [ [1, 2], [1, 3], [2, 3], [4, 5] ]
        >>> vertex_frame = tc.frame.create(vertex_rows, vertex_schema)
        >>> edge_frame = tc.frame.create(edge_rows, edge_schema)

        >>> graph = tc.graph.create(vertex_frame, edge_frame)

        >>> result = graph.connected_components()
        >>> result.inspect() 
        [#]  id  component
        ==================
        [0]   1          1
        [1]   2          1
        [2]   3          1
        [3]   4          4
        [4]   5          4

    """
    from sparktk.frame.frame import Frame
    return Frame(self._tc, self._scala.connectedComponents())
